File: dotnet/Crucible/Crucible-Runtime/Nodes/Node.cs
namespace Crucible.Nodes;
public class Node
{
    public Guid Uuid { get; private set; }
    public string Name { get; set; }
    
    private List<Node> _children = new List<Node>();
    public IReadOnlyCollection<Node> Children
    {
        get { return _children; }
    }
    private Node? _parent = null;

    public Node? Parent
    {
        get
        {
            return _parent;
        }
        set
        {
            if (_parent != null)
            {
                _parent._children.Remove(this);
            }
            _parent = value;
            if (_parent != null)
            {
                _parent._children.Add(this);
            }
        }
    }

    public Node()
    {
        Name = "Game Object";
        Uuid = Guid.NewGuid();
        GameWorld.nodes.Add(Uuid,new WeakReference<Node>(this));
    }

    private Node(string name, Guid uuid)
    {
        Name = name;
        Uuid = uuid;
    }

    ~Node()
    {
        GameWorld.nodes.Remove(Uuid);
    }

    public void AddChild(Node node)
    {
        node.Parent = this;
    }

    public override string ToString()
    {
        return Name;
    }
}

File: src/crucible/Core/Node.cpp
#include "Node.h"
#include <boost/uuid/uuid_generators.hpp>

namespace crucible
{

    boost::uuids::random_generator NODE_UUID_GENERATOR;

    Node::Node()
    {
        _id = NODE_UUID_GENERATOR();
    }

    Node::~Node()
    {
        std::remove(_parent->_children.begin(), _parent->_children.end(),this);
        for(Node* child: _children)
        {
            delete child;
        }
    }

    Node::Node(Node &&other)
    {
        move(std::move(other));
    }

    Node &Node::operator=(Node &&other)
    {
        move(std::move(other));
        return *this;
    }

    Node *Node::parent()
    {
        return _parent;
    }

    void Node::setParent(Node *node)
    {

        if(_parent)
        {
            std::remove(_parent->_children.begin(), _parent->_children.end(),this);
        }
        _parent = node;
        if(_parent)
        {
            _parent->_children.push_back(this);
        }
    }

    void Node::addChild(Node *node)
    {
        node->setParent(this);
    }

    void Node::removeChild(Node *node)
    {
        for(auto i=0; i< _children.size(); i++)
        {
            Node* comp = _children[i];
            if(comp == node)
            {
                removeChild(i);
                return;
            }
        }
    }

    void Node::removeChild(size_t childIndex)
    {
        Node* child = _children[childIndex];
        _children.erase(_children.begin()+childIndex);
        delete child;
    }

    Node *Node::getChild(size_t childIndex) const
    {
        return _children[childIndex];
    }

    size_t Node::childrenCount()const
    {
        return _children.size();
    }

    boost::uuids::uuid Node::id() const
    {
        return _id;
    }
} // crucible